Expose deprecation warning messages and stacktraces via build operations

breskeby opened the pull request:
### Context
exposing deprecation warnings + stacktrace allows build scan plugin to capture this data and ultimately make it available in gradle build scans.

We expose the plain stacktrace. it is only created when requested from the build operation api. therefore the overhead when no plugin is applied should be low.
  
### Contributor Checklist
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

ldaley commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecationWarningBuildOperation.java:
```diff
@@ -0,0 +1,55 @@
+
+                @Override
+                public List<StackTraceElement> getStackTrace() {
+                    return usage.withStackTrace().getStack();
```
This is going to lead to traces to the call of this method, and effectively makes this type mutable. Check the implementation of withStackTrace().

blindpirate replied:
+1

ldaley commented on subprojects/logging/src/integTest/groovy/org/gradle/DeprecationMessageBuildOperationIntegrationTest.groovy:
```diff
@@ -0,0 +1,47 @@
+        succeeds "t"
+
+        then:
+        def op = operations.first(DeprecationWarningBuildOperationType) { it.details.message.contains('test deprecation warning') }
```
We should be testing that the parent op is the task here. We should also be testing some other kinds of parents.

breskeby replied:
fixed. testing deprecations coming from different contexts now

ldaley commented on subprojects/logging/src/integTest/groovy/org/gradle/DeprecationMessageBuildOperationIntegrationTest.groovy:
```diff
@@ -0,0 +1,47 @@
+        def op = operations.first(DeprecationWarningBuildOperationType) { it.details.message.contains('test deprecation warning') }
+        op.details.stackTrace.size > 0
+        op.details.stackTrace[0].fileName.endsWith('build.gradle')
+        op.details.stackTrace[0].lineNumber == 4
```
This would not work if this deprecation warning were coming from buildSrc, because of the record/replay mechanism. 

We need to add some coverage for that kind of early (pre build scan) scenario, and verify that our test fixtures aren't giving us false positives.

breskeby replied:
added coverage for buildSrc and composites too

ldaley commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecationWarningBuildOperationType.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.List;
+
+public class DeprecationWarningBuildOperationType implements BuildOperationType<DeprecationWarningBuildOperationType.Details, DeprecationWarningBuildOperationType.Result> {
```
We should be modelling these things as progress events, instead of operations.

breskeby replied:
fixed

ldaley commented on subprojects/logging/src/integTest/groovy/org/gradle/DeprecationMessageBuildOperationIntegrationTest.groovy:
```diff
@@ -0,0 +1,47 @@
+
+    def operations = new BuildOperationsFixture(executer, temporaryFolder)
+
+    def "emits operation for deprecation warnings"() {
```
We need tests to verify that we get events for all usages, not just the initial.

blindpirate commented on subprojects/core/src/main/java/org/gradle/invocation/DefaultGradle.java:
```diff
@@ -94,11 +94,15 @@ public void projectsLoaded(Gradle gradle) {
             }
         });
 
+
```
I guess these empty lines can be removed.

breskeby replied:
fixed

breskeby commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:
```diff
@@ -68,11 +71,16 @@ public void featureUsed(FeatureUsage usage) {
                 LOGGER.warn(message.toString());
             }
         }
+        fireDeprecationWarningBuildOperation(usage);
```
@blindpirate that's on purpose. we want to track all deprecations. Further filtering will be done serverside

breskeby commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java:
```diff
@@ -120,7 +120,7 @@ private void executeAction(final String actionDisplayName, final TaskInternal ta
         buildOperationExecutor.run(new RunnableBuildOperation() {
             @Override
             public BuildOperationDescriptor.Builder description() {
-                return BuildOperationDescriptor.displayName(actionDisplayName + " for " + task.getPath()).name(actionDisplayName);
```
fixed this as stumbling upon this when evaluating task execution build operation events for composite builds

breskeby commented:
@adammurdoch changed the implementation to be based on progress events

ldaley commented on subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java:
```diff
@@ -163,7 +166,13 @@ private DefaultGradleLauncher doNewInstance(BuildDefinition buildDefinition,
+
+        BuildOperationListenerManager buildOperationListenerManager = serviceRegistry.get(BuildOperationListenerManager.class);
+        BuildOperationListener buildOperationBroadcaster= buildOperationListenerManager.getBroadcaster();
+        DeprecatedUsageBuildOperationProgressBroadaster deprecationWarningBuildOperationProgressBroadaster = new DeprecatedUsageBuildOperationProgressBroadaster(clock, buildOperationBroadcaster);
```
Seems like this could come from the service registry.

ldaley commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecatedUsageBuildOperationProgressBroadaster.java:
```diff
@@ -0,0 +1,64 @@
+
+public class DeprecatedUsageBuildOperationProgressBroadaster {
+
+    private final static OperationIdentifierProvider DEFAULT_CURRENT_OPERATION_IDENTIFIER = new OperationIdentifierProvider() {
```
You can just use `CurrentBuildOperationRef` instead of this.

breskeby replied:
I've added this mechanic to make the testing easier

ldaley replied:
How is it easier than just using `CurrentBuildOperationRef`?

ldaley commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecatedUsageProgressDetails.java:
```diff
@@ -0,0 +1,28 @@
+@UsedByScanPlugin
+public interface DeprecatedUsageProgressDetails {
+    String getMessage();
+    List<StackTraceElement> getStackTrace();
```
Where are the rest of the properties?

blindpirate commented on subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java:
```diff
@@ -59,102 +59,105 @@ public synchronized static void reportSuppressedDeprecations() {
+    /**
+     * Try to avoid using this nagging method. The other methods use a consistent wording for when things will be removed.
+     */
+    static void nagUserWith(String message) {
```
Why is this package-private? If this method is removed on purpose, why not private? However, my personal preference would be keeping it public like other overloads in case some plugin authors are using it - they usually have no idea `org.gradle.util` is not a public package.

ldaley replied:
We should not keep it because users might use it. This class is very much not public, and we need to be able to change it.

I like the user empathy, but in this case would should not speculatively keep it.

breskeby replied:
I kept it package private as its used only in some tests atm. I'll add a comment about this

oehme replied:
You can use `@VisibleForTesting` to express that

blindpirate commented on subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:
```diff
@@ -43,39 +43,47 @@
+
+    private void fireDeprecatedUsageBuildOperationProgress(FeatureUsage usage) {
+        if (buildOperationProgressBroadaster != null) {
+            buildOperationProgressBroadaster.progress(usage, usage.withStackTrace().getStack());
```
Can we make the stacktrace evaluation lazy? In my understanding, calculating stack trace is a relatively slow operation. My local experiment shows that invoking `FeatureUsage.withStacktrace().getStack()` can be as slow as 0.2ms with a 300-length stacktraces.

ldaley replied:
We can if we use `Exception` under the hood to get the trace, because it has some internal mechanics for delaying generation of the trace.

@breskeby we should try this.

oehme commented:
The model looks good to me.

We don't have performance test coverage for deprecation messages. We could create a small synthetic project that just exercises deprecation warnings, similar to the project that exercises just task creation.

If we want to get the stacktraces for every single usage then that's rather expensive, as @blindpirate already noted. We should get those on the background thread, not on the thread that creates the deprecation warning. I.e. create an Exception object and access its stacktrace later.

breskeby commented:
I'll look into adding some basic performance test coverage for this and then look into what's possible for improving based on suggestions by @oehme

breskeby commented:
I've added a dedicated performance test and changed the implementation to calculate the stack trace from the deprecation progress build operation events lazy.

oehme commented on subprojects/performance/templates.gradle:
```diff
@@ -80,6 +80,21 @@ tasks.register("createLotsOfTasks", JvmProjectGeneratorTask) {
+tasks.register("createLotsOfDeprecationWarnings", JvmProjectGeneratorTask) {
+    projects = 1
+    sourceFiles = 0
+    subProjectTemplates = ['deprecations']
```
You might wanna use a simple Copy task, like I did for the archivePerformanceTest. We don't need any of the advanced project generation stuff here.

breskeby commented:
> On Jul 12, 2018, at 8:42 PM, Stefan Oehme <notifications@github.com> wrote:
> 
> @oehme commented on this pull request.
> 
> In subprojects/logging/src/main/java/org/gradle/util/SingleMessageLogger.java <https://github.com/gradle/gradle/pull/5881#discussion_r202139073>:
> 
> >          }
>      }
>  
>      public static void nagUserOfReplacedNamedParameter(String parameterName, String replacement) {
>          if (isEnabled()) {
> -            nagUserWith(String.format(
> -                    "The %s named parameter %s. Please use the %s named parameter instead.",
> -                    parameterName, getDeprecationMessage(), replacement));
> +            nagUserWith(String.format("The %s named parameter has been deprecated.", parameterName), thisWillBeRemovedMessage(), String.format("Please use the %s named parameter instead.", replacement));
> +        }
> +    }
> +
> +    /**
> +     * Try to avoid using this nagging method. The other methods use a consistent wording for when things will be removed.
> +     */
> +    static void nagUserWith(String message) {
> You can use @VisibleForTesting to express that
> 
> 
Added. thought about it but its more like @ExistsForTesting. Anyhow added it.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/gradle/gradle/pull/5881#discussion_r202139073>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAEt9MuiFVVUvcH26KXpRVwvPyehj27Aks5uF5h4gaJpZM4VFGyI>.
>

breskeby commented:
With all the changes we made we now see a regression in the new introduced `DeprecationCreationPerformanceTest` of ~7 percent. I don’t see how we completely negate the overhead of our changes in BT. The question is, if we’re willing to accept the regression in this scenario. It’s not a typical build scenario and the currently 7 percent in a pure 10000 depreciations build seems kind of reasonable to me. It's not mine or @ldaley call to make though. @oehme wdyt?

ldaley commented:
The percentage amount is irrelevant as this build is not doing any work that's relative to the feature. The more useful metric is that 10,000 deprecations cost about 70ms (on the perf test hardware). 

I think this is an acceptable cost.

Having the deprecation traces in scans will be very valuable to users. It's likely to make them more visible (in big builds the current logging can easily be missed), easier to quantify, and easier to resolve (we can do more with the trace visualisation).

breskeby commented:
@oehme would be great if you can provide some feedback if you think this is good to go to be merged into master.

oehme commented:
Deprecations getting just 7% more expensive is a good result. For a real world build you'd probably struggle to ever measure it.

This is good to go from my side.

oehme commented:
Make sure to `@Ignore` that failing test until we have a nightly containing your changes that we can set as the baseline.
